From 4e850191efa6659de85a4ff9be7c896cfb0bb1ec Mon Sep 17 00:00:00 2001
Date: Mon, 17 May 2021 21:28:40 -0700
Subject: [PATCH 05/24] clang: support macports-libstdcxx and macports-libc++

requires reinplace of search paths
===
 clang/include/clang/Basic/DiagnosticDriverKinds.td
 clang/include/clang/Driver/ToolChain.h        |  3 +-
 clang/lib/Driver/ToolChain.cpp                |  3 ++
 clang/lib/Driver/ToolChains/Darwin.cpp        | 40 ++++++++++++++-----
 clang/lib/Driver/ToolChains/Gnu.cpp
 clang/lib/Driver/ToolChains/Hexagon.cpp       |  2 +
 clang/lib/Lex/InitHeaderSearch.cpp       |  2 +-

diff --git a/clang/include/clang/Basic/DiagnosticDriverKinds.td b/clang/include/clang/Basic/DiagnosticDriverKinds.td
index 1b69324..cf840b1 100644
--- a/clang/include/clang/Basic/DiagnosticDriverKinds.td
+++ b/clang/include/clang/Basic/DiagnosticDriverKinds.td
@@ -648,6 +648,10 @@ def warn_drv_libstdcxx_not_found : Warning<
   "command line to use the libc++ standard library instead">,
   InGroup<DiagGroup<"stdlibcxx-not-found">>;
 
+def err_drv_mplibcxx_not_found : Error<
+  "include path for macports-libc++ headers not found; pass '-stdlib=libc++' on the "
+  "command line to use the libc++ standard library instead">;
+
 def err_drv_cannot_mix_options : Error<"cannot specify '%1' along with '%0'">;
 
 def err_drv_invalid_object_mode : Error<
diff --git a/clang/include/clang/Driver/ToolChain.h b/clang/include/clang/Driver/ToolChain.h
index 2e74507..78b8ea9 100644
--- a/clang/include/clang/Driver/ToolChain.h
+++ b/clang/include/clang/Driver/ToolChain.h
@@ -94,6 +94,8 @@ public:
 
   enum CXXStdlibType {
     CST_Libcxx,
+    CST_MacPortsLibcxx,
+    CST_MacPortsLibstdcxx,
     CST_Libstdcxx
   };
 
diff --git a/clang/lib/Driver/ToolChain.cpp b/clang/lib/Driver/ToolChain.cpp
index 8dafc3d..bae5475 100644
--- a/clang/lib/Driver/ToolChain.cpp
+++ b/clang/lib/Driver/ToolChain.cpp
@@ -1070,6 +1070,10 @@ ToolChain::CXXStdlibType ToolChain::GetCXXStdlibType(const ArgList &Args) const{
     cxxStdlibType = ToolChain::CST_Libcxx;
   else if (LibName == "libstdc++")
     cxxStdlibType = ToolChain::CST_Libstdcxx;
+  else if (LibName == "macports-libc++" || LibName == "libc++_macports")
+    cxxStdlibType = ToolChain::CST_MacPortsLibcxx;
+  else if (LibName == "macports-libstdc++" || LibName == "libstdc++_macports")
+    cxxStdlibType = ToolChain::CST_MacPortsLibstdcxx;
   else if (LibName == "platform")
     cxxStdlibType = GetDefaultCXXStdlibType();
   else {
@@ -1199,6 +1203,13 @@ void ToolChain::AddCXXStdlibLibArgs(const ArgList &Args,
   CXXStdlibType Type = GetCXXStdlibType(Args);
 
   switch (Type) {
+  case ToolChain::CST_MacPortsLibcxx:
+    CmdArgs.push_back("-L@@MACPORTS_LIBCXX_DIR@@");
+#ifndef __APPLE__
+    CmdArgs.push_back("-rpath");
+    CmdArgs.push_back("@@MACPORTS_LIBCXX_DIR@@");
+#endif
+    // fallthrough to regular -stdlib=libc++ case:
   case ToolChain::CST_Libcxx:
     CmdArgs.push_back("-lc++");
     if (Args.hasArg(options::OPT_fexperimental_library))
@@ -1208,6 +1219,16 @@ void ToolChain::AddCXXStdlibLibArgs(const ArgList &Args,
   case ToolChain::CST_Libstdcxx:
     CmdArgs.push_back("-lstdc++");
     break;
+  case ToolChain::CST_MacPortsLibstdcxx:
+#ifdef __APPLE__
+    CmdArgs.push_back("-lstdc++");
+#else
+    // CmdArgs.push_back("-lstdc++"); // this is too unpredictable
+    CmdArgs.push_back("-rpath");
+    CmdArgs.push_back("@@MACPORTS_GCC_LIB_DIR@@");
+    CmdArgs.push_back("@@MACPORTS_GCC_LIB_DIR@@/libstdc++.so"); // full path to runtime so we're certain we get the right one
+#endif
+    break;
   }
 }
 
diff --git a/clang/lib/Driver/ToolChains/Darwin.cpp b/clang/lib/Driver/ToolChains/Darwin.cpp
index 7da2a83..a886fb5 100644
--- a/clang/lib/Driver/ToolChains/Darwin.cpp
+++ b/clang/lib/Driver/ToolChains/Darwin.cpp
@@ -2512,6 +2512,27 @@ void DarwinClang::AddClangCXXStdlibIncludeArgs(
     break;
   }
 
+  case ToolChain::CST_MacPortsLibcxx: {
+    llvm::SmallString<128> UsrIncludeCxx("@@MACPORTS_LIBCXX_INCLUDE_DIR@@");
+    if (getVFS().exists(UsrIncludeCxx)) {
+      addSystemInclude(DriverArgs, CC1Args, UsrIncludeCxx);
+      return;
+    } else {
+      getDriver().Diag(diag::err_drv_mplibcxx_not_found);
+    }
+    break;
+  }
+
+  case ToolChain::CST_MacPortsLibstdcxx: {
+    bool IsBaseFoundMacPorts = AddGnuCPlusPlusIncludePaths(DriverArgs, CC1Args, llvm::StringRef("@@MACPORTS_GCC_INCLUDE_DIR@@"),
+						   "",
+						   "@@MACPORTS_HOST_NAME@@",
+						   @@MACPORTS_TEST_32_64@@);
+    if (!IsBaseFoundMacPorts) {
+      getDriver().Diag(diag::warn_drv_libstdcxx_not_found);
+    }}
+    break;
+
   case ToolChain::CST_Libstdcxx:
     llvm::SmallString<128> UsrIncludeCxx = Sysroot;
     llvm::sys::path::append(UsrIncludeCxx, "usr", "include", "c++");
@@ -2565,12 +2586,19 @@ void DarwinClang::AddCXXStdlibLibArgs(const ArgList &Args,
   CXXStdlibType Type = GetCXXStdlibType(Args);
 
   switch (Type) {
+  case ToolChain::CST_MacPortsLibcxx:
+    CmdArgs.push_back("-L@@MACPORTS_LIBCXX_DIR@@");
+    // fallthrough to regular -stdlib=libc++ case:
   case ToolChain::CST_Libcxx:
     CmdArgs.push_back("-lc++");
     if (Args.hasArg(options::OPT_fexperimental_library))
       CmdArgs.push_back("-lc++experimental");
     break;
 
+  case ToolChain::CST_MacPortsLibstdcxx:
+    CmdArgs.push_back("@@MACPORTS_libstdc++@@");
+    break;
+
   case ToolChain::CST_Libstdcxx:
     // Unfortunately, -lstdc++ doesn't always exist in the standard search path;
     // it was previously found in the gcc lib dir. However, for all the Darwin
diff --git a/clang/lib/Driver/ToolChains/Gnu.cpp b/clang/lib/Driver/ToolChains/Gnu.cpp
index 40038dc..a1cfbf9 100644
--- a/clang/lib/Driver/ToolChains/Gnu.cpp
+++ b/clang/lib/Driver/ToolChains/Gnu.cpp
@@ -3094,6 +3094,14 @@ void Generic_GCC::AddClangCXXStdlibIncludeArgs(const ArgList &DriverArgs,
   case ToolChain::CST_Libstdcxx:
     addLibStdCxxIncludePaths(DriverArgs, CC1Args);
     break;
+  case ToolChain::CST_MacPortsLibcxx:
+    addSystemInclude(DriverArgs, CC1Args, "@@MACPORTS_LIBCXX_INCLUDE_DIR@@");
+    break;
+  case ToolChain::CST_MacPortsLibstdcxx:
+    addSystemInclude(DriverArgs, CC1Args, "@@MACPORTS_GCC_INCLUDE_DIR@@");
+    addSystemInclude(DriverArgs, CC1Args, "@@MACPORTS_GCC_INCLUDE_DIR@@/@@MACPORTS_HOST_NAME@@");
+    addSystemInclude(DriverArgs, CC1Args, "@@MACPORTS_GCC_INCLUDE_DIR@@/backward");
+    break;
   }
 }
 
diff --git a/clang/lib/Driver/ToolChains/Hexagon.cpp b/clang/lib/Driver/ToolChains/Hexagon.cpp
index aed4ab1..e0f8fad 100644
--- a/clang/lib/Driver/ToolChains/Hexagon.cpp
+++ b/clang/lib/Driver/ToolChains/Hexagon.cpp
@@ -767,6 +767,10 @@ HexagonToolChain::GetCXXStdlibType(const ArgList &Args) const {
       return ToolChain::CST_Libstdcxx;
   }
   StringRef Value = A->getValue();
+  if (Value == "macports-libc++" || Value == "libc++_macports")
+    return ToolChain::CST_MacPortsLibcxx;
+  if (Value == "macports-libstdc++" || Value == "libstdc++_macports")
+    return ToolChain::CST_MacPortsLibstdcxx;
   if (Value != "libstdc++" && Value != "libc++")
     getDriver().Diag(diag::err_drv_invalid_stdlib_name) << A->getAsString(Args);
 
diff --git a/clang/lib/Lex/InitHeaderSearch.cpp b/clang/lib/Lex/InitHeaderSearch.cpp
index 41382d7..27ecaa8 100644
--- a/clang/lib/Lex/InitHeaderSearch.cpp
+++ b/clang/lib/Lex/InitHeaderSearch.cpp
@@ -116,7 +116,7 @@ static bool CanPrefixSysroot(StringRef Path) {
 #if defined(_WIN32)
   return !Path.empty() && llvm::sys::path::is_separator(Path[0]);
 #else
-  return llvm::sys::path::is_absolute(Path);
+  return llvm::sys::path::is_absolute(Path) && (Path.find("@@MACPORTS_GCC_INCLUDE_DIR@@")!=0 || Path.find("@@MACPORTS_LIBCXX_INCLUDE_DIR@@")!=0);
 #endif
 }
 
